test/common/PtrUnion.cpp

```diff
@@ -0,0 +1,354 @@
+    case ncclFloat32: return 4;
+    case ncclFloat64: return 8;
+    case ncclBfloat16: return 2;
+    default:
+      ERROR("Unsupported datatype (%d)\n", dataType);
+      exit(0);
+    }
+  }
+
+  ErrCode PtrUnion::Attach(void *ptr)
+  {
+    this->ptr = ptr;
+    return TEST_SUCCESS;
+  }
+
+  ErrCode PtrUnion::Attach(PtrUnion ptrUnion)
+  {
+    this->ptr = ptrUnion.ptr;
+    return TEST_SUCCESS;
+  }
+
+  ErrCode PtrUnion::AllocateGpuMem(size_t const numBytes, bool const useManagedMem)
+  {
+    if (numBytes)
+    {
+      if (useManagedMem)
+      {
+        if (hipMallocManaged(&I1, numBytes) != hipSuccess)
+        {
+          ERROR("Unable to allocate managed memory of GPU memory (%lu bytes)\n", numBytes);
+          return TEST_FAIL;
+        }
+      }
+      else
+      {
+        if (hipMalloc(&I1, numBytes) != hipSuccess)
+        {
+          ERROR("Unable to allocate memory of GPU memory (%lu bytes)\n", numBytes);
+          return TEST_FAIL;
+        }
+      }
+    }
+    return TEST_SUCCESS;
+  }
+
+  ErrCode PtrUnion::AllocateCpuMem(size_t const numBytes)
+  {
+    if (numBytes)
+    {
+      this->ptr = calloc(numBytes, 1);
+      if (!ptr)
+      {
+        ERROR("Unable to allocate memory (%lu bytes)\n", numBytes);
+        return TEST_FAIL;
+      }
+    }
+    return TEST_SUCCESS;
+  }
+
+  ErrCode PtrUnion::FreeGpuMem()
+  {
+    if (this->ptr != nullptr)
+    {
+      hipFree(this->ptr);
+      this->ptr = nullptr;
+    }
+    return TEST_SUCCESS;
+  }
+
+  ErrCode PtrUnion::FreeCpuMem()
+  {
+    if (this->ptr != nullptr)
+    {
+      free(this->ptr);
+      this->ptr = nullptr;
+    }
+    return TEST_SUCCESS;
+  }
+
+  ErrCode PtrUnion::ClearGpuMem(size_t const numBytes)
+  {
+    if (hipMemset(this->ptr, 0, numBytes) != hipSuccess)
+    {
+      ERROR("Unable to call hipMemset\n");
+      return TEST_FAIL;
+    }
+    return TEST_SUCCESS;
+  }
+
+  ErrCode PtrUnion::ClearCpuMem(size_t const numBytes)
+  {
+    memset(this->ptr, 0, numBytes);
+    return TEST_SUCCESS;
+  }
+
+  ErrCode PtrUnion::FillPattern(ncclDataType_t const dataType,
+                                size_t         const numElements,
+                                int            const globalRank,
+                                bool           const isGpuMem)
+  {
+    PtrUnion temp;
+    size_t const numBytes = numElements * DataTypeToBytes(dataType);
+
+    // If this is GPU memory, create a CPU temp buffer otherwise fill CPU memory directly
+    if (isGpuMem)
+      temp.AllocateCpuMem(numBytes);
+    else
+      temp.Attach(this->ptr);
+
+    for (int i = 0; i < numElements; i++)
+    {
+      int    valueI = (globalRank + i) % 256;
+      double valueF = 1.0L/((double)valueI+1.0L);
+      temp.Set(dataType, i, valueI, valueF);
+    }
+
+    // If this is GPU memory, copy from CPU temp buffer
+    if (isGpuMem)
+    {
+      if (hipMemcpy(this->ptr, temp.ptr, numBytes, hipMemcpyHostToDevice) != hipSuccess)
```

Comment:
could replace if statement here with CHECK_HIP(hipMemset(this->ptr, 0, numBytes)) for consistency through the tests
that would also need a comment here with the error stated underneath though